fs: fix `fs.rm` support for loop symlinks

nathanael-ruf opened the pull request:
Fixes: https://github.com/nodejs/node/issues/45404

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

nathanael-ruf commented:
The tests have quite a bit of duplicated code (both existing and my new tests). I can add a new block at the bottom of the file that adds the new tests for the sync, async and promise version without so much duplicated code if you want.

Trott commented:
> The tests have quite a bit of duplicated code (both existing and my new tests). I can add a new block at the bottom of the file that adds the new tests for the sync, async and promise version without so much duplicated code if you want.

For tests only, I generally prefer the repetition. (Ref: https://stackoverflow.com/a/11837973/436641)

aduh95 commented on test/parallel/test-fs-rm.js:
```diff
@@ -49,6 +49,15 @@ function makeNonEmptyDirectory(depth, files, folders, dirname, createSymLinks) {
+        path.join(dirname, `link-${depth}-loop-${y}`),
+        'file'
+      );
+    })
```
```suggestion
    });
```

nathanael-ruf replied:
Thanks, fixed.

aduh95 commented on test/parallel/test-fs-rm.js:
```diff
@@ -137,6 +146,48 @@ function removeAsync(dir) {
+  // Should delete an invalid symlink
+  const invalidLink = path.join(tmpdir.path, 'invalid-link-async');
+  fs.symlinkSync('definitely-does-not-exist-async', invalidLink);
+  fs.rm(validLink, common.mustNotMutateObjectDeep({ recursive: true }), common.mustCall((err) => {
```
I think you meant `invalidLink` here. Isolating each test in a code block could help catch this kind of typos.
```suggestion
  fs.rm(invalidLink, common.mustNotMutateObjectDeep({ recursive: true }), common.mustCall((err) => {
```

nathanael-ruf replied:
Yes, thanks for catching it. Fixed.

> Isolating each test in a code block could help catch this kind of typos.

Agree.

nathanael-ruf commented:
Thanks for fixing the commit message, one question though: For something like addressing the review comments, should I always amend and force push or is another commit okay for this (I guess you can always squash at the end)?

Also I think the title of the PR is not longer 100% accurate, because deleting invalid links was also broken before (the `stat` call failed with `ENOENT`):

```
nathanael ~/tmp λ tree
.
└── b -> a

nathanael ~/tmp λ node -e 'fs.rmSync("b")'
node:internal/fs/utils:344
    throw err;
    ^

Error: ENOENT: no such file or directory, stat 'b'
    at Object.statSync (node:fs:1538:3)
    at __node_internal_ (node:internal/fs/utils:793:8)
    at Object.rmSync (node:fs:1213:13)
    at [eval]:1:4
    at Script.runInThisContext (node:vm:129:12)
    at Object.runInThisContext (node:vm:305:38)
    at node:internal/process/execution:76:19
    at [eval]-wrapper:6:22
    at evalScript (node:internal/process/execution:75:60)
    at node:internal/main/eval_string:27:3 {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: 'b'
}
```

The loop thing was just the first issue I encountered.

aduh95 commented:
> should I always amend and force push or is another commit okay for this (I guess you can always squash at the end)?

Pushing more commits is the recommended way, it's a much better experience for folks reviewing the PR as otherwise GitHub is not able to show what's changed since the last review. When landing the PR, we can ask the commit-queue to squash all the commits into the first one, so as long as its commit message is accurate, no need to force push. The other reason to force push would be if you needed to rebase on top of `main`, e.g. to fix a git conflict (merge commits tend to break our tooling).

> On 13 Nov 2022, at 09:04, Nathanael Ruf ***@***.***> wrote:
> 
> 
> Thanks for fixing the commit message, one question though: For something like addressing the review comments, should I always amend and force push or is another commit okay for this (I guess you can always squash at the end)?
> 
> Also I think the title of the PR is not longer 100% accurate, because deleting invalid links was also broken before (the stat call failed with ENOENT):
> 
> nathanael ~/tmp λ tree
> .
> └── b -> a
> 
> nathanael ~/tmp λ node -e 'fs.rmSync("b")'
> node:internal/fs/utils:344
>     throw err;
>     ^
> 
> Error: ENOENT: no such file or directory, stat 'b'
>     at Object.statSync (node:fs:1538:3)
>     at __node_internal_ (node:internal/fs/utils:793:8)
>     at Object.rmSync (node:fs:1213:13)
>     at [eval]:1:4
>     at Script.runInThisContext (node:vm:129:12)
>     at Object.runInThisContext (node:vm:305:38)
>     at node:internal/process/execution:76:19
>     at [eval]-wrapper:6:22
>     at evalScript (node:internal/process/execution:75:60)
>     at node:internal/main/eval_string:27:3 {
>   errno: -2,
>   syscall: 'stat',
>   code: 'ENOENT',
>   path: 'b'
> }
> The loop thing was just the first issue I encountered.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/nodejs/node/pull/45439#issuecomment-1312665509>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADNFTDN6WSRGOPCR5JPUCXDWICOKVANCNFSM6AAAAAAR6O64FQ>.
> You are receiving this because you commented.
>

aduh95 commented:
/cc @nodejs/fs

LiviaMedeiros commented on test/parallel/test-fs-rm.js:
```diff
@@ -181,6 +232,39 @@ if (isGitPresent) {
+  const validLink = path.join(tmpdir.path, 'valid-link');
+  fs.symlinkSync(linkTarget, validLink);
+  try {
+    fs.rmSync(validLink);
```
This code doesn't seem to test if `fs.rmSync` succeeds.
```suggestion
    fs.rmSync(validLink);
    assert.strictEqual(fs.existsSync(validLink), false);
```

nathanael-ruf replied:
Thanks, done.

LiviaMedeiros commented on test/parallel/test-fs-rm.js:
```diff
@@ -181,6 +232,39 @@ if (isGitPresent) {
+  const invalidLink = path.join(tmpdir.path, 'invalid-link');
+  fs.symlinkSync('definitely-does-not-exist', invalidLink);
+  try {
+    fs.rmSync(invalidLink);
```
Same here
```suggestion
    fs.rmSync(invalidLink);
    assert.strictEqual(fs.existsSync(invalidLink), false);
```

nathanael-ruf replied:
Thanks, done.

LiviaMedeiros commented on test/parallel/test-fs-rm.js:
```diff
@@ -181,6 +232,39 @@ if (isGitPresent) {
+  fs.symlinkSync(loopLinkA, loopLinkB);
+  fs.symlinkSync(loopLinkB, loopLinkA);
+  try {
+    fs.rmSync(loopLinkA);
```
Same here, and optionally can be fixed in old code
```suggestion
    fs.rmSync(loopLinkA);
    assert.strictEqual(fs.existsSync(loopLinkA), false);
```

nathanael-ruf replied:
Thanks, done.

I also fixed in existing code even though it's not strictly related to the issue this PR is fixing. If that's not okay I can open a separate PR for these changes instead, no problem.

I also just want to note that this would probably not have happened if this would've been shared code between the callback, promise and sync version tests, because someone clearly thought of this for the callback version but forgot to make the same change in two other places.

LiviaMedeiros replied:
Thank you. I think, fixing those in this PR is fine.

Historically, `fs` tests tend to be repetitive because Promises API came much later than others, and usually combining sync/awaited code with Callback API in one wrapper function is not trivial. Many tests are even split into 3 separate files to avoid complexity and potential race conditions.
In general (e.g. in new tests) I agree with your suggested approach; but IMHO refactoring this particular test might make it less readable.

nathanael-ruf replied:
Alright, thanks for the explanation.

LiviaMedeiros reviewed (commented):
LGTM with a suggestion.

nathanael-ruf commented:
Looks like `existsSync` (and therefore `access`) returns false on an invalid symlink (see my last commit). Is this also a bug or intended behaviour?

LiviaMedeiros commented on test/parallel/test-fs-rm.js:
```diff
@@ -177,25 +228,64 @@ if (isGitPresent) {
+  const loopLinkB = path.join(tmpdir.path, 'loop-link-b');
+  fs.symlinkSync(loopLinkA, loopLinkB);
+  fs.symlinkSync(loopLinkB, loopLinkA);
+  try {
+    fs.rmSync(loopLinkA);
+  } finally {
+    assert.strictEqual(fs.existsSync(loopLinkA), false);
```
```suggestion
  try {
    fs.rmSync(loopLinkA);
    assert.strictEqual(fs.existsSync(loopLinkA), false);
  } finally {
```

LiviaMedeiros commented:
> Looks like `existsSync` (and therefore `access`) returns false on an invalid symlink (see my last commit). Is this also a bug or intended behaviour?

I think this is intended. In general, symlinks are supposed to be transparent, so accessing invalid symlink should be equal to accessing its destination (e.g. nonexistent file).

LiviaMedeiros commented:
Thanks for the contribution!
